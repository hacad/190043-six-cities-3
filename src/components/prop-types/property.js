import PropTypes from "prop-types";
import CityPropType from "./city.js";
import LocationPropType from "./location.js";

const PropertyPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  city: CityPropType.isRequired,
  type: PropTypes.string.isRequired,
  previewImage: PropTypes.string.isRequired,
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
  isPremium: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  rating: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  goods: PropTypes.arrayOf(PropTypes.string).isRequired,
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    isPro: PropTypes.bool.isRequired,
    avatarUrl: PropTypes.string.isRequired
  }),
  location: LocationPropType
});

export default PropertyPropType;
