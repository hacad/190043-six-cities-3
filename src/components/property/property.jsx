import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import PropertyPropType from "../prop-types/property.js";
import ReviewPropType from "../prop-types/review.js";
import PlacePropType from "../prop-types/place.js";
import {ActionCreator, Operation} from "../../reducers/data/reducer";
import ReducerNames from "../../reducers/reducer-names.js";
import ReviewList from "../review-list/review-list.jsx";
import {getOfferById, getComments, getSelectedPlaces} from "../../reducers/data/selectors.js";
import Header from "../header/header.jsx";
import BookmarkButton from "../bookmark-button/bookmark-button.jsx";
import withAuthorization from "../../hocs/with-authorization/with-authorization.js";
import CitiesMap from "../cities-map/cities-map.jsx";
import PlacesList from "../places-list/places-list.jsx";

const HeaderWrapped = withAuthorization(Header);
const BookmarkButtonWrapped = withAuthorization(BookmarkButton);
const ReviewListWrapped = withAuthorization(ReviewList);

class Property extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.loadComments();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.offer.id !== this.props.offer.id) {
      this.props.loadComments();
    }
  }

  render() {
    const {offer, comments, nearOffers, hotelId, activeOffer, onActivateItem, onDeactivateItem} = this.props;

    if (!offer) {
      return null;
    }

    const activeCitiesMapOffer = activeOffer
      ? {
        data: activeOffer.location,
        displaySettings: {
          icon: {
            iconUrl: `/img/pin-active.svg`
          }
        }
      }
      : undefined;

    const nearOfferLocations = nearOffers.map((nearOffer) => {
      return {data: nearOffer.location};
    });
    nearOfferLocations.push({
      data: offer.location,
      displaySettings: {
        icon: {
          iconUrl: `/img/pin-active.svg`
        }
      }
    });

    return (
      <div className="page">
        <HeaderWrapped />
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {
                  offer.images.map((src) => {
                    return (
                      <div key={`property__image-${src}`} className="property__image-wrapper">
                        <img className="property__image" src={src} alt="Photo studio" />
                      </div>
                    );
                  })
                }
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {offer.isPremium
                  ? <div className="property__mark">
                    <span>Premium</span>
                  </div>
                  : ``
                }
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {offer.title}
                  </h1>
                  <BookmarkButtonWrapped
                    offerId={offer.id}
                    isFavorite={offer.isFavorite}
                    iconWidth={31}
                    iconHeight={33}
                    classNamePrefix="property"
                  />
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${offer.starRating}%`}}></span>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{offer.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    Entire place
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {offer.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {offer.maxAdults} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{offer.price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {
                      offer.goods.map((good) => {
                        return (
                          <li key={`property__inside-item-${good}`} className="property__inside-item">
                            {good}
                          </li>
                        );
                      })
                    }
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={`/${offer.host.avatarUrl}`} width="74" height="74" alt="Host avatar" />
                    </div>
                    <span className="property__user-name">
                      {offer.host.name}
                    </span>
                    <span className="property__user-status">
                      {offer.host.isPro ? `Pro` : ``}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {offer.description}
                    </p>
                  </div>
                </div>
                <ReviewListWrapped hotelId={hotelId} reviews={comments} />
              </div>
            </div>
            <CitiesMap city={offer.city} offers={nearOfferLocations} activeOffer={activeCitiesMapOffer} className="property__map map"/>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <PlacesList
                places={nearOffers}
                onClickCardHeader={() => {}}
                className="near-places__list places__list"
                onActivatePlace={onActivateItem}
                onDeactivatePlace={onDeactivateItem}
                articleTagClassNamePrefix="near-places__card"
                divImageWrapperClassNamePrefix="near-places"
              />
            </section>
          </div>
        </main>
      </div>
    );
  }
}

Property.propTypes = {
  hotelId: PropTypes.number.isRequired,
  offer: PropertyPropType,
  nearOffers: PropTypes.arrayOf(PlacePropType),
  comments: PropTypes.arrayOf(ReviewPropType),
  loadComments: PropTypes.func.isRequired,
  activeOffer: PlacePropType,
  onActivateItem: PropTypes.func.isRequired,
  onDeactivateItem: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => {
  const hotelId = parseInt(ownProps.match.params.id, 10);

  return Object.assign({}, ownProps, {
    hotelId,
    offer: getOfferById(hotelId, state),
    nearOffers: getSelectedPlaces(state)
                .filter((offer) => offer.id !== hotelId)
                .slice(0, 3),
    comments: getComments(state),
    activeOffer: state[ReducerNames.DATA].activeOffer
  });
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    loadComments: () => {
      dispatch(Operation.loadComments(parseInt(ownProps.match.params.id, 10)));
    },
    onActivateItem: (activeOffer) => {
      dispatch(ActionCreator.changeActiveOffer(activeOffer));
    },
    onDeactivateItem: () => {
      dispatch(ActionCreator.changeActiveOffer(undefined));
    }
  };
};

export {Property};

export default connect(mapStateToProps, mapDispatchToProps)(Property);
