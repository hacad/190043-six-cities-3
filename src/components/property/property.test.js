import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter as Router} from "react-router-dom";
import {Property} from "./property.jsx";

jest.mock(`../../hocs/with-authorization/with-authorization.js`, () => () => `<div />`);

const testProperty = {
  id: 1,
  city: {
    name: `Amsterdam`,
    location: {
      latitude: 52.370216,
      longitude: 4.895168,
      zoom: 10
    }
  },
  type: `apartment`,
  previewImage: `img/apartment-01.jpg`,
  images: [`1.jpg`, `2.jpg`, `3.jpg`],
  isPremium: false,
  title: `Beautiful &amp; luxurious apartment at great location`,
  price: 120,
  rating: 93,
  starRating: 93,
  isFavorite: false,
  goods: [`Coffee machine`, `Wi-Fi`],
  host: {
    id: 25,
    name: `Angelina`,
    isPro: true,
    avatarUrl: `img/avatar-angelina.jpg`
  },
  location: {
    latitude: 52.370216,
    longitude: 4.895168
  }
};

const reviews = [
  {
    "id": 1,
    "user": {
      "id": 18,
      "isPro": true,
      "name": `Sophie`,
      "avatarUrl": `https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/9.jpg`
    },
    "rating": 4,
    "comment": `The deluxe room was a quite comfortable one with all the adequate facilities. The only thing that made me feel uncomfortable was the rude behavior of an impolite staff at the reception desk.`,
    "date": new Date(`2019-12-03T14:11:47.471Z`)
  }
];

it(`Property correctly renders after relaunch`, () => {
  const offer = testProperty;

  const property = renderer.create(
      <Router>
        <Property
          hotelId={1}
          offer={offer}
          nearOffers={[]}
          comments={reviews}
          loadComments={jest.fn()}
          className="near-places__list places__list"
          handleItemActivate={jest.fn()}
          handleItemDeactivate={jest.fn()} />
      </Router>).toJSON();

  expect(property).toMatchSnapshot();
});
