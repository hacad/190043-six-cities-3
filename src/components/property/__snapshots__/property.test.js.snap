// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property correctly renders after relaunch 1`] = `
<div
  className="page"
>
  <<div /> />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="1.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="2.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="3.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful &amp; luxurious apartment at great location
            </h1>
            <button
              className="property__bookmark-button button "
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon property__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "93%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              93
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              Entire place
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              120
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Wi-Fi
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="/img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              />
            </div>
          </div>
          <<div />
            hotelId={1}
            reviews={
              Array [
                Object {
                  "comment": "The deluxe room was a quite comfortable one with all the adequate facilities. The only thing that made me feel uncomfortable was the rude behavior of an impolite staff at the reception desk.",
                  "date": 2019-12-03T14:11:47.471Z,
                  "id": 1,
                  "rating": 4,
                  "user": Object {
                    "avatarUrl": "https://htmlacademy-react-2.appspot.com/six-cities/static/avatar/9.jpg",
                    "id": 18,
                    "isPro": true,
                    "name": "Sophie",
                  },
                },
              ]
            }
          />
        </div>
      </div>
    </section>
    <section
      className="property__map map"
      id="map"
    />
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        />
      </section>
    </div>
  </main>
</div>
`;
